Author: Mort65, https://github.com/mort65/
Description: This patch toggles the visibility of all tags on current monitor
diff -up /home/mohsen/dwm/config.def.h ./config.def.h
--- /home/mohsen/dwm/config.def.h	2019-02-07 21:05:32.382956919 +0330
+++ ./config.def.h	2019-02-07 21:08:19.919621503 +0330
@@ -64,6 +64,7 @@ static Key keys[] = {
 	{ MODKEY,                       XK_p,      spawn,          {.v = dmenucmd } },
 	{ MODKEY|ShiftMask,             XK_Return, spawn,          {.v = termcmd } },
 	{ MODKEY,                       XK_b,      togglebar,      {0} },
+	{ MODKEY,                       XK_z,      toggletags,      {0} },
 	{ MODKEY,                       XK_j,      focusstack,     {.i = +1 } },
 	{ MODKEY,                       XK_k,      focusstack,     {.i = -1 } },
 	{ MODKEY,                       XK_i,      incnmaster,     {.i = +1 } },
diff -up /home/mohsen/dwm/dwm.c ./dwm.c
--- /home/mohsen/dwm/dwm.c	2019-02-07 21:05:32.396290253 +0330
+++ ./dwm.c	2019-02-07 21:18:27.969613944 +0330
@@ -130,6 +130,8 @@ struct Monitor {
 	Monitor *next;
 	Window barwin;
 	const Layout *lt[2];
+	unsigned int curtagset[2];
+	int showtags;
 };

 typedef struct {
@@ -213,6 +215,9 @@ static void togglebar(const Arg *arg);
 static void togglefloating(const Arg *arg);
 static void toggletag(const Arg *arg);
 static void toggleview(const Arg *arg);
+static void toggletags(const Arg *arg);
+static void showtags(const Arg *arg);
+static void hidetags(const Arg *arg);
 static void unfocus(Client *c, int setfocus);
 static void unmanage(Client *c, int destroyed);
 static void unmapnotify(XEvent *e);
@@ -634,6 +639,8 @@ createmon(void)

 	m = ecalloc(1, sizeof(Monitor));
 	m->tagset[0] = m->tagset[1] = 1;
+	m->curtagset[0] = m->curtagset[1] = 1;
+	m->showtags = 1;
 	m->mfact = mfact;
 	m->nmaster = nmaster;
 	m->showbar = showbar;
@@ -1049,6 +1056,10 @@ manage(Window w, XWindowAttributes *wa)
 		&& (c->x + (c->w / 2) < c->mon->wx + c->mon->ww)) ? bh : c->mon->my);
 	c->bw = borderpx;

+	if (!c->tags)
+		c->tags = c->mon->showtags ? 1 : c->mon->curtagset[c->mon->seltags];
+
 	wc.border_width = c->bw;
 	XConfigureWindow(dpy, w, CWBorderWidth, &wc);
 	XSetWindowBorder(dpy, w, scheme[SchemeNorm][ColBorder].pixel);
@@ -1737,6 +1748,9 @@ toggletag(const Arg *arg)
 void
 toggleview(const Arg *arg)
 {
+	if (!selmon->showtags)
+		showtags(NULL);
+
 	unsigned int newtagset = selmon->tagset[selmon->seltags] ^ (arg->ui & TAGMASK);

 	if (newtagset) {
@@ -1747,6 +1761,32 @@ toggleview(const Arg *arg)
 }

 void
+toggletags(const Arg *arg) {
+	selmon->showtags ? hidetags(NULL) : showtags(NULL);
+}
+
+void
+hidetags(const Arg *arg)
+{
+	for (int i = 0; i < LENGTH(selmon->curtagset); i++)
+		selmon->curtagset[i] = selmon->tagset[i];
+	selmon->tagset[selmon->seltags] = 0;
+	selmon->showtags = 0;
+	focus(NULL);
+	arrange(selmon);
+}
+
+void
+showtags(const Arg *arg)
+{
+	for (int i = 0; i < LENGTH(selmon->curtagset); i++)
+		selmon->tagset[i] |= selmon->curtagset[i];
+	selmon->showtags = 1;
+	focus(NULL);
+	arrange(selmon);
+}
+
+void
 unfocus(Client *c, int setfocus)
 {
 	if (!c)
@@ -2040,6 +2080,8 @@ view(const Arg *arg)
 	selmon->seltags ^= 1; /* toggle sel tagset */
 	if (arg->ui & TAGMASK)
 		selmon->tagset[selmon->seltags] = arg->ui & TAGMASK;
+	if (!selmon->showtags)
+		selmon->showtags = 1;
 	focus(NULL);
 	arrange(selmon);
 }
